src/_Derivatives/subexpressions.jl

```diff
@@ -14,78 +14,100 @@ function list_subexpressions(nd::Vector{NodeData})
+    # exist) are added, so we're free to append subexpression `n` to the list.
+    push!(individual[1], n)
+    push!(individual[2], n)
+    # In addition, if we are storing the full list, add `n` here too.
+    if full !== nothing && !(n in full[2])
+        push!(full[1], n)
+        push!(full[2], n)
     end
+    return
+end
+
+"""
+    order_subexpressions(
+        main_expression::Vector{NodeData},
+        subexpressions::Vector{Vector{NodeData}},
+        full::Union{Nothing,Tuple{Vector{Int64},Set{Int64}}} = nothing,
+    )
+
+Return a topologically ordered list of subexpression-indices needed to evaluate
+`main_expression`.
+
+`full` is an argument used by `order_subexpressions` to collate a list of
+expressions needed to evaluate multiple `main_expressions`.
 
-    # Generate an individual order for each main expression.
-    individual_order = [Int[] for i in 1:length(main_expressions)]
-    for o in order_filtered
-        for i in depended_on_by[o]
-            push!(individual_order[i], o)
+**Warning:** This doesn't handle cyclic expressions! But this should be fine
+because we can't compute them in JuMP anyway.
+"""
+function order_subexpressions(
+    main_expression::Vector{NodeData},
+    subexpressions::Vector{Vector{NodeData}},
+    full::Union{Nothing,Tuple{Vector{Int64},Set{Int64}}} = nothing,
+)
+    # `order` is a topologically sorted list of nodes. Nodes at the start of the
+    # list need to be evaluated first.
+    # `tagged` is a set to efficiently keep track of nodes that we have added to
+    # `order`. We could just use `n in order`, but a `Set` is probably faster.
+    order, tagged = Int64[], Set{Int64}()
+    # The meat of DFS is this recursive call to `_visit_node` for each
+    # SUBEXPRESSION node in `main_expression`.
+    for node in main_expression
+        if node.nodetype == SUBEXPRESSION
+            _visit_nodes(subexpressions, node.index, (order, tagged), full)
         end
     end
+    return order
+end
+
+"""
+    order_subexpressions(
+        main_expressions::Vector{Vector{NodeData}},
+        subexpressions::Vector{Vector{NodeData}};
+    )
+
+Topologically sort the subexpressions needed to evaluate `main_expressions`.
 
-    return order_filtered, individual_order
+Returns two things:
+
+ * A `Vector{Int64}` containing the ordered list of subexpression-indices that
+   need to be evaluated to compute all `main_expressions`
+ * A `Vector{Vector{Int64}}`, containing a list of ordered lists of
+   subexpression-indices that need to be evaluated to compute
+   `main_expressions[i]`.
+
+**Warning:** This doesn't handle cyclic expressions! But this should be fine
+because we can't compute them in JuMP anyway.
+"""
+function order_subexpressions(
+    main_expressions::Vector{Vector{NodeData}},
+    subexpressions::Vector{Vector{NodeData}},
+)
+    full = (Int64[], Set{Int64}())
+    individual = Vector{Int64}[
+        order_subexpressions(expr, subexpressions, full)
```

Comment:
The complexity of this is O(sum(# of nodes in each depended-on subexpression)), because we scan the subexpressions each time we see them.